File: src/main/java/etlmail/engine/css/SelectorSpecificity.java
package etlmail.engine.css;

class SelectorSpecificity implements Comparable<SelectorSpecificity> {
	private final int ids;
	private final int classes;
	private final int types;

	public SelectorSpecificity(int ids, int classes, int types) {
		this.ids = ids;
		this.classes = classes;
		this.types = types;
	}

	public static class Builder {
		private int ids;
		private int classes;
		private int types;

		public Builder addId(int count) {
			ids += count;
			return this;
		}

		public Builder addClass(int count) {
			classes += count;
			return this;
		}

		public Builder addType(int count) {
			types += count;
			return this;
		}

		public SelectorSpecificity asSpecificity() {
			return new SelectorSpecificity(ids, classes, types);
		}
	}

	@Override
	public int compareTo(SelectorSpecificity o) {
		int result = ids - o.ids;
		if (result == 0) {
			result = classes - o.classes;
		}
		if (result == 0) {
			result = types - o.types;
		}
		return result;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = (prime * result) + classes;
		result = (prime * result) + ids;
		result = (prime * result) + types;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (getClass() != obj.getClass()) {
			return false;
		}
		final SelectorSpecificity other = (SelectorSpecificity) obj;
		if (classes != other.classes) {
			return false;
		}
		if (ids != other.ids) {
			return false;
		}
		if (types != other.types) {
			return false;
		}
		return true;
	}

	@Override
	public String toString() {
		return "SelectorSpecificit(" + ids + ", " + classes + ", " + types
				+ ")";
	}

}

File: src/main/scala/etlmail/engine/css/SelectorSpecificity.scala
package etlmail.engine.css;

class SelectorSpecificity(val ids: Int, val classes: Int, val types: Int) extends Ordered[SelectorSpecificity] {

  override def compare(o: SelectorSpecificity): Int = {
    var result = ids - o.ids;
    if (result == 0) {
      result = classes - o.classes;
    }
    if (result == 0) {
      result = types - o.types;
    }
    return result;
  }

  override def hashCode(): Int = {
    val prime = 31;
    var result = 1;
    result = (prime * result) + classes;
    result = (prime * result) + ids;
    result = (prime * result) + types;
    return result;
  }

  override def equals(obj: Any): Boolean =
    (this eq obj.asInstanceOf[AnyRef]) || (obj match {
      case null => false
      case other: SelectorSpecificity =>
        classes == other.classes && ids == other.ids && types == other.types
      case _ => false
    })

  override def toString(): String = "SelectorSpecificit(" + ids + ", " + classes + ", " + types + ")"
}

object SelectorSpecificity {
  class Builder {
    private var ids: Int = 0
    private var classes: Int = 0
    private var types: Int = 0

    def addId(count: Int): Builder = {
      ids += count;
      return this;
    }

    def addClass(count: Int): Builder = {
      classes += count;
      return this;
    }

    def addType(count: Int): Builder = {
      types += count;
      return this;
    }

    def asSpecificity(): SelectorSpecificity = new SelectorSpecificity(ids, classes, types)
  }
}